projects/br_scs/src/Mirza/SupplyChain/Handlers/EventRegistration.hs

```diff
@@ -16,172 +14,34 @@ import           Mirza.SupplyChain.Database.Schema as Schema
-                                                    OutputEPC (..),
-                                                    ParentLabel (..),
-                                                    TransactionDWhat (..),
-                                                    TransformationDWhat (..))
+import           Data.GS1.DWhat                    (LabelEPC (..))
 import           Data.GS1.Event                    as Ev
 
-import           Data.Maybe                        (isJust)
-
-import           Control.Monad.Except              (when)
-
-insertObjectEvent :: (Member context '[HasDB],
+insertGS1Event :: (Member context '[HasDB],
                       Member err     '[AsSqlError])
-                  => ObjectEvent
+                  => Ev.Event
                   -> AppM context err (EventInfo, Schema.EventId)
-insertObjectEvent ob = runDb $ insertObjectEventQuery ob
-
-insertObjectEventQuery :: ObjectEvent
-                       -> DB context err (EventInfo, Schema.EventId)
-insertObjectEventQuery
-  (ObjectEvent
-    foreignEventId
-    act
-    labelEpcs
-    dwhen dwhy dwhere
-  ) = do
-  let
-      dwhat =  ObjWhat $ ObjectDWhat act labelEpcs
-      event = Ev.Event Ev.ObjectEventT foreignEventId dwhat dwhen dwhy dwhere
-
-  -- insertEvent has to be the first thing that happens here so that
-  -- uniqueness of the JSON event is enforced
-  (evInfo, eventId) <- insertEvent event
-  whatId <- insertDWhat Nothing dwhat eventId
-  labelIds' <- mapM insertLabel labelEpcs
-  let labelIds = Schema.LabelId <$> labelIds'
-  _whenId <- insertDWhen dwhen eventId
-  _whyId <- insertDWhy dwhy eventId
-  insertDWhere dwhere eventId
-  mapM_ (insertWhatLabel Nothing (Schema.WhatId whatId)) labelIds
-  mapM_ (insertLabelEvent Nothing eventId) labelIds
-  pure (evInfo, eventId)
-
-
-insertAggEvent  :: (Member context '[HasDB],
-                    Member err     '[AsSqlError])
-                => AggregationEvent
-                -> AppM context err (EventInfo, Schema.EventId)
-insertAggEvent ev = runDb $ insertAggEventQuery ev
-
-insertAggEventQuery :: AggregationEvent
-                    -> DB context err (EventInfo, Schema.EventId)
-insertAggEventQuery
-  (AggregationEvent
-    foreignEventId
-    act
-    mParentLabel
-    labelEpcs
-    dwhen dwhy dwhere
-  ) = do
-  let
-      dwhat =  AggWhat $ AggregationDWhat act mParentLabel labelEpcs
-      event = Ev.Event Ev.AggregationEventT foreignEventId dwhat dwhen dwhy dwhere
+insertGS1Event ev = runDb $ insertEventQuery ev
 
+insertEventQuery :: Ev.Event
+                 -> DB context err (EventInfo, Schema.EventId)
+insertEventQuery
+  event@(Ev.Event _eType _foreignEventId dwhat dwhen dwhy dwhere) = do
   -- insertEvent has to be the first thing that happens here so that
   -- uniqueness of the JSON event is enforced
   (evInfo, eventId) <- insertEvent event
   whatId <- insertDWhat Nothing dwhat eventId
+  let (labelWithTypes :: [LabelWithType]) = getLabelsWithType dwhat
```

Comment:
+3. Why the explicit type signatures here? getLabelsWithType is monomorphic, GHC has no problems resolving the types.

You can also drop ScopedTypeVariables after the change.